Framework/Core/include/Framework/TableTreeHelpers.h

```diff
@@ -22,97 +22,205 @@ namespace o2::framework
-  arrow::Field* mField;
-  arrow::Type::type mFieldType;
-  arrow::Type::type mElementType;
-  int32_t mNumberElements;
-  std::string mLeaflistString;
-
-  TBranch* mBranchPtr = nullptr;
-
-  char* mBranchBuffer = nullptr;
-  void* mValueBuffer = nullptr;
-
-  std::shared_ptr<arrow::BooleanArray> mArray_o = nullptr;
-  //bool mBoolValueHolder;
-  bool* mVariable_o = nullptr;
-
-  uint8_t* mVariable_ub = nullptr;
-  uint16_t* mVariable_us = nullptr;
-  uint32_t* mVariable_ui = nullptr;
-  uint64_t* mVariable_ul = nullptr;
-  int8_t* mVariable_b = nullptr;
-  int16_t* mVariable_s = nullptr;
-  int32_t* mVariable_i = nullptr;
-  int64_t* mVariable_l = nullptr;
-  float* mVariable_f = nullptr;
-  double* mVariable_d = nullptr;
-
-  // initialize a branch
-  bool initBranch(TTree* tree);
-
-  // initialize chunk ib
-  bool initDataBuffer(Int_t ib);
+ protected:
+  std::string mBranchName;
 
- public:
-  BranchIterator(TTree* tree, std::shared_ptr<arrow::ChunkedArray> col, std::shared_ptr<arrow::Field> field);
-  ~BranchIterator();
+  virtual void resetBuffer() = 0;
+  virtual void nextChunk() = 0;
 
-  // has the iterator been properly initialized
-  bool getStatus();
+  arrow::ChunkedArray* mColumn = nullptr;
 
-  // fills buffer with next value
-  // returns false if end of buffer reached
-  bool push();
+  int64_t const* mCurrentPos = nullptr;
+  mutable int mFirstIndex = 0;
+  mutable int mCurrentChunk = 0;
+
+  int listSize = 1;
 };
 
-class TableToTree
-{
+template <typename T>
+struct ROOTTypeString {
+  static constexpr char const* str = "/E";
+};
 
- private:
-  TTree* mTreePtr;
+template <>
+struct ROOTTypeString<bool> {
+  static constexpr char const* str = "/O";
+};
 
-  // a list of BranchIterator
-  std::vector<std::unique_ptr<BranchIterator>> mBranchIterators;
+template <>
+struct ROOTTypeString<uint8_t> {
+  static constexpr char const* str = "/b";
+};
 
-  // table to convert
-  std::shared_ptr<arrow::Table> mTable;
+template <>
+struct ROOTTypeString<uint16_t> {
+  static constexpr char const* str = "/s";
+};
 
+template <>
+struct ROOTTypeString<uint32_t> {
+  static constexpr char const* str = "/i";
+};
+
+template <>
+struct ROOTTypeString<uint64_t> {
+  static constexpr char const* str = "/l";
+};
+
+template <>
+struct ROOTTypeString<int8_t> {
+  static constexpr char const* str = "/B";
+};
+
+template <>
+struct ROOTTypeString<int16_t> {
+  static constexpr char const* str = "/S";
+};
+
+template <>
+struct ROOTTypeString<int32_t> {
+  static constexpr char const* str = "/I";
+};
+
+template <>
+struct ROOTTypeString<int64_t> {
+  static constexpr char const* str = "/L";
+};
+
+template <>
+struct ROOTTypeString<float> {
+  static constexpr char const* str = "/F";
+};
+
+template <>
+struct ROOTTypeString<double> {
+  static constexpr char const* str = "/D";
+};
+
+template <typename T>
+constexpr auto ROOTTypeString_t = ROOTTypeString<T>::str;
+
+template <typename T>
+class ColumnToBranch : public ColumnToBranchBase
```

Comment:
Can't you convert T to a string and remove the need for templating?

Reply:
I need the template to define the actual storing pointer, or at least cast it to correct type.